### ISSUE TITLE
Designing the Neural Network for the RL Agent

### ISSUE THREAD

**Issue by PaulZbigniew** [KaufmannLukas/ds-ultimate-tic-tac-toe#14]
This happens after we defined the base structure for the agent.

---

_**NN SETUP**_

**Input representation:**
How to represent the game state as input to the neural network?
=> use CNN to capture spacial relationships within the board

**Network architecture**
Which architecture to choose for the NN?
i.e. deep neural networks, such as CNN
=> experiment with different network architectures and layer sizes to find the best 
=> our network should also be able to process the game state and produce Q-values for different values

**Output layer**
=> the output layer of the chosen NN should have as many nodes as there are possible actions in the game
=> each node in the output layer corresponds to a different action the agent can take
=> the network should produce Q-values. They represent the expected future rewards for taking each action in the current state

**Activation functions**
Which activation function to choose for hidden layers?
i.e. ReLU or variants like Leaky ReLU

**Loss function**
Which loss function to choose?
i.e. MSE loss for Q-learning

**Training Procedure**
=> Train the NN (using data from self-play)

**Exploration vs. Exploitation**
Which exploration strategy to implement?
i.e. epsilon-greedy
=> Need to balance exploration (exploring new actions) vs. exploitation (choosing best-known actions) 
=> crucial for a robust agent

---

**_TUNING_**

**Regularization & Hyperparameters**
=> experiment with regularization techniques to prevent overfitting
=> tune hyperparameters 

**Iterate and Experiment**
=> train, evaluate, and adjust the network architecture and parameters

**Evaluate and Monitor**
=> monitor the agent's performance (does it make progress?)
=> evaluate and test against different opponents or strategies
i.e. let generation n play against generation n-1 or n-2 etc.
